### ISSUE TITLE
Canvas ruins the FPS in Mali GPU

### ISSUE THREAD

**Issue by knowledgenude** [armory3d/armory#1695]
The issue is detailed here in the last comment: https://github.com/armory3d/armory/issues/1693

**Comment by knowledgenude** [comment]
I do tests with more elements:

![canvasissue](https://user-images.githubusercontent.com/63247726/82091812-6dee8900-96ce-11ea-92a1-865cd909c7aa.jpg)

Test file (contains exported APKs):

[canvas_test.zip](https://github.com/armory3d/armory/files/4636928/canvas_test.zip)

**Comment by QuantumCoderQC** [comment]
@knowledgenude 

Thanks for more information.

Could you also elaborate on the method you used to test. 

And did you use the Debug version or the Release version. Both versions have been known to show different results. A comparison of Debug vs Release versions may also help. 

I also want to know if the app was built specifically for the device on Android Studio, or built for one device and then installed on the other. 

Best Regards, 
QC

**Comment by knowledgenude** [comment]
I don't targetted a specific phone, but i have tested Android Studio with my phone connected to it sometimes, i don't know if this stay saved in some place.

The build variant is debug x86. All the phones i have tested are 64 bits phones and i don't think this could affect the gpu performance but i can test with all settings you want of course.

Now i will try to export the apk as signed because just changing the build to release, give me an unsigned apk that i can't install in any of the phones. I always get the message "the app was not installed" when i try it.

Here is all the settings i have used in Armory and Android Studio:

[prints.zip](https://github.com/armory3d/armory/files/4637314/prints.zip)

**Comment by knowledgenude** [comment]
I don't understand why the Canvas element are bigger in the mali device. Both devices have the same resolution (1920x1080) but in the mali one, the elements are bigger. Unfortunately don't have two mali devices from  same resolution to compare this aspect but this is not having influence in the performance. Can you tell  me what the Canvas takes in count to set its size? Because even the displays are from the same inches (5.5 in both).

**Comment by knowledgenude** [comment]
The issue is on released signed apk too. Maybe using apk bundle can solve this. I will try to target the especific device but i don't have too much hope on this :/

**Comment by onelsonic** [comment]
Hi @knowledgenude what your phones GPU specs?
It might be inherent to the Mali GPU itself not being powerful enough.
https://www.notebookcheck.net/Smartphone-Graphics-Cards-Benchmark-List.149363.0.html

**Comment by knowledgenude** [comment]
@onelsonic , i don't think it is fault of the hardware, because without canvas elements (using only geometry) there is no lag. Also i noticed that textured objects are not working in mobile and back in time there is poor performance on textured objects (only in this mali gpu). Considering that this devices runs full games and apps with a lot of textures (including alphas) and objects, i can't affirm that is a hardware fault nor a canvas specific fault.

The phone spec is:

Octa-core 1.6 GHz Cortex-A53

Mali-T830 MP1

I have almost sure if you take a high end device with a mali gpu it will lag too.

**Comment by onelsonic** [comment]
@knowledgenude 

so one phone is a 
Octa-core 1.6 GHz Cortex-A53
Mali-T830 MP1

what is the other phone specs? and what target platform are you testing? Android, WebGL2, Krom ?

I am currently testing a project I made with armory with a 
Mali-G71   MP2
and an
Qualcomm Adreno 540

strangely the Mail is undertaking the Adreano in fragment shaders task for webgl 2 for example.

So you might need to drill down to GPU specs comparing apple to apples on the same platform first in order to find what is wrong.

**Comment by e1e5en-gd** [comment]
I did a Render Path comparison on different devices, specs and results here (Android target platform):
https://forums.armory3d.org/t/koui-user-interface-library-for-armory-and-kha/4087/37?u=e1e5en

**Comment by onelsonic** [comment]
if you render it on android (openglES)
can you for both phones publish the Devise specs and OpenGL features supported :
https://play.google.com/store/apps/details?id=com.realtechvr.glview

and if you're using the html5 webGL1/2 the specs can be found here :
webGL2
https://webglreport.com/?v=2

webGL1 (Legacy Shaders)
https://webglreport.com/?v=1
